### ISSUE TITLE
SQLite 3.37 STRICT TABLE support

### ISSUE THREAD

**Issue by NeiroNet** [ErikEJ/SqlCeToolbox#921]
Hi!
Version SQLite 3.37 introduces support for STRICT tables.
Can I add this?

**Comment by ErikEJ** [comment]
Is there a System.data
Sqlite release that includes this?

**Comment by ErikEJ** [comment]
Looks like it is planned for this month: https://system.data.sqlite.org/index.html/doc/trunk/www/news.wiki

**Comment by ErikEJ** [comment]
@NeiroNet Finally a new version was released!

**Comment by ErikEJ** [comment]
Meh. Unable to load .116 in VS 2022 - seems like VS is hardcoded to use .109 from the PrivateAssemblies folder.

@madskristensen any idea why that is? (Talking about System.Data.Sqlite dll)

**Comment by madskristensen** [comment]
You may need to add ProvideCodeBase to your assembly referencing the dll. @aarnott does that sound right?

**Comment by ErikEJ** [comment]
Yeah, looks promising! I will give it a try.

**Comment by AArnott** [comment]
You should generally *not* ship dll's in your extension that VS already ships. And you should generally keep your dependencies to be no newer than the copies that VS installs. To do anything else is inviting pain.

So ideally, rollback your dependency to .109. If you *must* use a newer version, and the binding redirects in VS don't force even newer version to match the older version that it ships, ProvideCodeBase should be able to help you. But don't expect type equivalence between the types in the assembly you ship and the types from the assembly that VS ships.

**Comment by ErikEJ** [comment]
I would really like to use .116, as it adds support for new SQLite engine features, including strict tables, but sadly it seems impossible. Wonder what MS tool is using this non-MS libraray, when a Microsoft alternative exists (Microsoft.Data.Sqlite) (yes, I know I use it, old story and some features still missing from the "replacement")

And it did not work:

**Redirect found in application configuration file: 1.0.116.0 redirected to 1.0.109.0**

`<description>SetSite failed for package [SqlCeToolboxPackage]Source: &apos;SqlCeVsToolbox&apos; Description: Could not load file or assembly &apos;System.Data.SQLite, Version=1.0.116.0, Culture=neutral, PublicKeyToken=db937bc2d44ff139&apos; or one of its dependencies. The system cannot find the file specified.&#x000D;&#x000A;System.IO.FileNotFoundException: Could not load file or assembly &apos;System.Data.SQLite, Version=1.0.116.0, Culture=neutral, PublicKeyToken=db937bc2d44ff139&apos; or one of its dependencies. The system cannot find the file specified.&#x000D;&#x000A;File name: &apos;System.Data.SQLite, Version=1.0.116.0, Culture=neutral, PublicKeyToken=db937bc2d44ff139&apos;&#x000D;&#x000A;   at ErikEJ.SqlCeToolbox.SqlCeToolboxPackage.&lt;InitializeAsync&gt;d__17.MoveNext()&#x000D;&#x000A;   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine](TStateMachine&amp; stateMachine)&#x000D;&#x000A;   at ErikEJ.SqlCeToolbox.SqlCeToolboxPackage.InitializeAsync(CancellationToken cancellationToken, IProgress`1 progress)&#x000D;&#x000A;   at Microsoft.VisualStudio.Shell.AsyncPackage.&lt;&gt;c__DisplayClass20_0.&lt;&lt;Microsoft-VisualStudio-Shell-Interop-IAsyncLoadablePackageInitialize-Initialize&gt;b__1&gt;d.MoveNext()&#x000D;&#x000A;--- End of stack trace from previous location where exception was thrown ---&#x000D;&#x000A;   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()&#x000D;&#x000A;   at Microsoft.VisualStudio.Services.VsTask.RethrowException(AggregateException e)&#x000D;&#x000A;   at Microsoft.VisualStudio.Services.VsTask.InternalGetResult(Boolean ignoreUIThreadCheck)&#x000D;&#x000A;&#x000D;&#x000A;=== Pre-bind state information ===&#x000D;&#x000A;LOG: DisplayName = System.Data.SQLite, Version=1.0.116.0, Culture=neutral, PublicKeyToken=db937bc2d44ff139&#x000A; (Fully-specified)&#x000D;&#x000A;LOG: Appbase = file:///C:/Program Files/Microsoft Visual Studio/2022/Enterprise/Common7/IDE/&#x000D;&#x000A;LOG: Initial PrivatePath = NULL&#x000D;&#x000A;Calling assembly : SqlCeVsToolbox, Version=4.7.210.0, Culture=neutral, PublicKeyToken=3681435cabd17ad2.&#x000D;&#x000A;===&#x000D;&#x000A;LOG: This bind starts in LoadFrom load context.&#x000D;&#x000A;WRN: Native image will not be probed in LoadFrom context. Native image will only be probed in default load context, like with Assembly.Load().&#x000D;&#x000A;LOG: Using application configuration file: C:\Users\Erik\AppData\Local\Microsoft\VisualStudio\17.0_a2ed33e9Exp\devenv.exe.config&#x000D;&#x000A;LOG: Using host configuration file: &#x000D;&#x000A;LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework64\v4.0.30319\config\machine.config.&#x000D;&#x000A;LOG: Redirect found in application configuration file: 1.0.116.0 redirected to 1.0.109.0.&#x000D;&#x000A;LOG: Using codebase from policy file: file:///C:/Program Files/Microsoft Visual Studio/2022/Enterprise/Common7/IDE/PrivateAssemblies/System.Data.SQLite.dll.&#x000D;&#x000A;LOG: Post-policy reference: System.Data.SQLite, Version=1.0.109.0, Culture=neutral, PublicKeyToken=db937bc2d44ff139&#x000D;&#x000A;LOG: Attempting download of new URL file:///C:/Program Files/Microsoft Visual Studio/2022/Enterprise/Common7/IDE/PrivateAssemblies/System.Data.SQLite.dll.&#x000D;&#x000A;WRN: Comparing the assembly name resulted in the mismatch: Build Number&#x000D;&#x000A;</description>
`

**Comment by ErikEJ** [comment]
And this is the binding redirect in devenv.exe.config:


```xml
        <dependentAssembly>
          <assemblyIdentity name="System.Data.SQLite" publicKeyToken="db937bc2d44ff139" culture="neutral"/>
          <bindingRedirect oldVersion="0.0.0.0-2.0.0.0" newVersion="1.0.109.0"/>
          <codeBase version="1.0.109.0" href="PrivateAssemblies\System.Data.SQLite.dll"/>
        </dependentAssembly>

```

**Comment by ErikEJ** [comment]
Yeah, yikes! My extension runs fine (using .109), but I am now at the mercy of MS in order to adopt new SQLite features - not good. 

Any fix in newer VS releases will be HIGHLY appreciated, then I can always point users needing STRICT tables to use the newer version.

@AArnott I really appreciate your support!

**Comment by AArnott** [comment]
So maybe some good news: that binding redirect was only added in 17.2. So if we fix it, you'll work in 17.1 and earlier, as well as some newer version with our fix.

**Comment by ErikEJ** [comment]
Ah! Sweet - assume I still need ProvideCodeBase (that currently crashes in 17.2) ?

`[assembly: ProvideCodeBase(AssemblyName = "System.Data.SQLite", Version = "1.0.116.0", Culture = "neutral", PublicKeyToken = "db937bc2d44ff139", CodeBase = "$PackageFolder$\\System.Data.SQLite.dll")]
`

**Comment by ErikEJ** [comment]
Would you like me to log an issue for this somewhere else than here?

**Comment by AArnott** [comment]
Yes, filing a Feedback Ticket will help add fuel to the fire to get this in a 17.2 servicing release (which _might_ be feasible). Can you share a link here when you have it so I can make sure it gets routed and linked to a PR with a fix I have already created?

**Comment by AArnott** [comment]
A `ProvideCodeBaseAttribute` is always a good idea for any/all assemblies that your extension ships.

**Comment by MaxG117** [comment]
I'm the author of the VS bug report.

@AArnott I started looking at using AppDomain last week. It’s a bit of a pain. I don’t think we can start a VS extension in a separate AppDomain, because we need the VSSDK libraries that are available in the default AppDomain. I could start my SQLite component in its own AppDomain, but ideally it should have access to the VS status bar, and I haven’t figured out how to pass it across the domain boundary (I'm new to AppDomain).

**Comment by AArnott** [comment]
@MaxG117 Don't plan on being able to bring *any* of the VS SDK into your appdomain. You'll have to do all VS interactions in the default appdomain and use your own only for sqlite operations. You'll have to design your own marshallable or serializable data objects for passing across the appdomain boundary.

**Comment by MaxG117** [comment]
@AArnott that makes sense. Would be nice to know what the VS team plans to do about this problem, if anything, before I invest any more time into a workaround.

**Comment by AArnott** [comment]
@MaxG117 That makes sense. I suggest you watch the [feedback ticket](https://developercommunity.visualstudio.com/t/Visual-Studio-extension-cant-use-System/10060849?entry=problem&ref=native&refTime=1654538978133&refUserId=4dfbe216-7a6c-41ab-ba73-494c825ac2a0) for updates.

**Comment by ErikEJ** [comment]
@AArnott Any updates/timeline?

**Comment by AArnott** [comment]
The fix has been checked in for 17.3 Preview 1 and later. I'm pushing the owning team to explore whether we could service 17.2 for this.

**Comment by ErikEJ** [comment]
Wonderful. I will test that version...

**Comment by AArnott** [comment]
I'm sorry. I misread the commit data. The fix _has_ been checked in for 17.3, but it isn't in any of the released previews yet.

**Comment by ErikEJ** [comment]
No problem. Looking forward to the next one a d will check the config file.

**Comment by ErikEJ** [comment]
@AArnott Looks like this was fixed in 17.2.6 - thanks so much for your help!

```xml
        <dependentAssembly>
          <assemblyIdentity name="System.Data.SQLite" publicKeyToken="db937bc2d44ff139" culture="neutral"/>
          <bindingRedirect oldVersion="0.0.0.0-1.0.109.0" newVersion="1.0.109.0"/>
          <codeBase version="1.0.109.0" href="PrivateAssemblies\System.Data.SQLite.dll"/>
        </dependentAssembly>

```

**Comment by ErikEJ** [comment]
Just works now in 17.2.6!

```sql
CREATE TABLE t1(a INT) STRICT;
INSERT INTO t1 VALUES('000123');
SELECT * FROM t1
INSERT INTO t1 VALUES('abc');

```
Result:

ErrorCode : 19
Message   : constraint failed
cannot store TEXT value in INT column t1.a
Result    : Constraint
